# Capstone Project - The Battle of the Neighborhoods (Week 2)

## Index
* [Introduction: Business Problem](#introduction)
* [Data](#data)
* [Methodology](#methodology)
* [Analysis](#analysis)
* [Results and Discussion](#results)
* [Conclusion](#conclusion)



## Introduction: Business Problem <a name="introduction"></a>

For this project I decided to explore the cities of Oakland, Emeryville, and San Diego to get a better glimpse into potential suggestions for investors and entrepreneurs looking to make a sound investment in restauranteering. I was looking for what type of restaurant would be the best and in which city to focus on when scouting for potential property.

The question we attempted to address by conducting this project was to discern how accurately we can predict the amount of "likes" a new restaurant opening in this region can expect to have based on the type of cuisine it will serve and which city in California it will open in. For this project I analyzed and modeled the data using machine learning by comparing both linear and logistic regressions to see which method yielded better predictive capabilities after training and testing.

## Data <a name="data"></a>

Once this is complete, we finally name the dataframe 'raw_dataset' as it is the most complete compiled form before needing any processing for analysis via machine learning.

First we retrieved the geographical coordinates of the three cities (Oakland, Emeryville, and San Diego). We then went ahead and leveraged the Foursquare API in order to obtain the URLs that gave us the raw data in JSON form. Each respective URL was then scraped for the columns: 'name', 'categories', 'latitude', 'longitude', and'id' for each city. The city column will help us when separating where the restaurants are from.

For this project, I decided to focus on those restaurants found within a 1000km radius from the coordinates that were provided by the geolocator. The Foursquare API provides us with more venue categories than we need, and therefore we had to make sure to clean our results by removing non-restaurant rows. Pulling the 'likes' data is necessary for us to make our final decisions. We don't want to be pulling information that will be discarded anyways and is of no valuable for our analysis.

We used the 'id' column in order to pull the 'likes' using the API and append the information into the dataframe. We concluded by naming the dataframe 'raw_dataset', which we used in the machine learning portion of the project.

## Methodology <a name="methodology"></a>

Both linear and logistic regression were used to train and test the data. Linear regression was used to predict the number of 'likes' a new restaurant in this region will acquire. Sci-Kit Learn was employed for this stage.

Logistic regression was used as the classification method. Since, we used binning when classifying by number of 'likes', we essentially made use of multinomial logistic regression to perform our analysis. Although the ranges are discrete categories, they can be considered ordinal in nature. The logistic regression will need to be specified as being both multinomial and ordinal. The Sci-Kit Learn package is perfect for this.

## Analysis <a name="analysis"></a>

#### **Importing Libraries**

In [1]:
import pandas as pd 
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
import numpy as np 
import json
import requests
from pandas.io.json import json_normalize 
import matplotlib.cm as cm
import matplotlib.colors as colors
import folium 
import matplotlib.pyplot as plt
import pylab as pl
import itertools
import warnings
warnings.filterwarnings('ignore')

from urllib.request import urlopen
from bs4 import BeautifulSoup
from geopy.geocoders import Nominatim 
from sklearn import linear_model
from sklearn.metrics import jaccard_score
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.metrics import log_loss
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error, r2_score

print('All libraries have been imported')

All libraries have been imported


#### **Retrieving Foursquare City Coordinates:**

In [2]:
address1 = 'Oakland, California'

geolocator = Nominatim(user_agent="foursquare_agent")
location1 = geolocator.geocode(address1)
latitude1 = location1.latitude
longitude1 = location1.longitude
print('The geograpical coordinate of {} are {}, {}.'.format(address1, latitude1, longitude1))

address2 = 'Emeryville, California'

geolocator = Nominatim(user_agent="foursquare_agent")
location2 = geolocator.geocode(address2)
latitude2 = location2.latitude
longitude2 = location2.longitude
print('The geograpical coordinate of {} are {}, {}.'.format(address2, latitude2, longitude2))

address3 = 'San Diego, California'

geolocator = Nominatim(user_agent="foursquare_agent")
location3 = geolocator.geocode(address3)
latitude3 = location3.latitude
longitude3 = location3.longitude
print('The geograpical coordinate of {} are {}, {}.'.format(address3, latitude3, longitude3))

The geograpical coordinate of Oakland, California are 37.8044557, -122.2713563.
The geograpical coordinate of Emeryville, California are 37.8314089, -122.2865266.
The geograpical coordinate of San Diego, California are 32.7174202, -117.1627728.


#### **Foursquare Credentials**

In [3]:
CLIENT_ID = 'QWYCDTT0ZVHE2FKH3EOAA0ZV2XQCOKFXOLWZ1QZ0AHIKRYTJ' #  Foursquare ID
CLIENT_SECRET = 'CGTSO3EP1A0OXCKSOXR0T2LQRONFYTHXRLIBPYRSQS5U22ZK' #  Foursquare Secret
VERSION = '20180604' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)


LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 1000 # define radius

# create URLs
url1 = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    latitude1, 
    longitude1, 
    radius, 
    LIMIT)


# create URLs
url2 = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    latitude2, 
    longitude2, 
    radius, 
    LIMIT)


# create URLs
url3 = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    latitude3, 
    longitude3, 
    radius, 
    LIMIT)

print(url1, url2, url3)

Your credentails:
CLIENT_ID: QWYCDTT0ZVHE2FKH3EOAA0ZV2XQCOKFXOLWZ1QZ0AHIKRYTJ
CLIENT_SECRET:CGTSO3EP1A0OXCKSOXR0T2LQRONFYTHXRLIBPYRSQS5U22ZK
https://api.foursquare.com/v2/venues/explore?&client_id=QWYCDTT0ZVHE2FKH3EOAA0ZV2XQCOKFXOLWZ1QZ0AHIKRYTJ&client_secret=CGTSO3EP1A0OXCKSOXR0T2LQRONFYTHXRLIBPYRSQS5U22ZK&v=20180604&ll=37.8044557,-122.2713563&radius=1000&limit=100 https://api.foursquare.com/v2/venues/explore?&client_id=QWYCDTT0ZVHE2FKH3EOAA0ZV2XQCOKFXOLWZ1QZ0AHIKRYTJ&client_secret=CGTSO3EP1A0OXCKSOXR0T2LQRONFYTHXRLIBPYRSQS5U22ZK&v=20180604&ll=37.8314089,-122.2865266&radius=1000&limit=100 https://api.foursquare.com/v2/venues/explore?&client_id=QWYCDTT0ZVHE2FKH3EOAA0ZV2XQCOKFXOLWZ1QZ0AHIKRYTJ&client_secret=CGTSO3EP1A0OXCKSOXR0T2LQRONFYTHXRLIBPYRSQS5U22ZK&v=20180604&ll=32.7174202,-117.1627728&radius=1000&limit=100


In [4]:
# scrape the data from the generated URLs

results1 = requests.get(url1).json()
results1

results2 = requests.get(url2).json()
results2

results3 = requests.get(url3).json()
results3

# function that extracts the category of the venue

def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']
    

# FIRST CITY   

venues1 = results1['response']['groups'][0]['items']
nearby_venues1 = pd.json_normalize(venues1) # flatten JSON

# filter columns
filtered_columns1 = ['venue.name', 'venue.categories', 'venue.location.lat', 
                    'venue.location.lng', 'venue.id']
nearby_venues1 = nearby_venues1.loc[:, filtered_columns1]

# filter the category for each row
nearby_venues1['venue.categories'] = nearby_venues1.apply(get_category_type, axis=1)

# clean columns
nearby_venues1.columns = [col.split(".")[-1] for col in nearby_venues1.columns]

# SECOND CITY

venues2 = results2['response']['groups'][0]['items']
nearby_venues2 = pd.json_normalize(venues2) # flatten JSON

# filter columns
filtered_columns2 = ['venue.name', 'venue.categories', 'venue.location.lat', 
                    'venue.location.lng', 'venue.id']
nearby_venues2 = nearby_venues2.loc[:, filtered_columns2]

# filter the category for each row
nearby_venues2['venue.categories'] = nearby_venues2.apply(get_category_type, axis=1)

# clean columns
nearby_venues2.columns = [col.split(".")[-1] for col in nearby_venues2.columns]

# THIRD CITY

venues3 = results3['response']['groups'][0]['items']
nearby_venues3 = pd.json_normalize(venues3) # flatten JSON

# filter columns
filtered_columns3 = ['venue.name', 'venue.categories', 'venue.location.lat', 
                    'venue.location.lng', 'venue.id']
nearby_venues3 = nearby_venues3.loc[:, filtered_columns3]

# filter the category for each row
nearby_venues3['venue.categories'] = nearby_venues3.apply(get_category_type, axis=1)

# clean columns
nearby_venues3.columns = [col.split(".")[-1] for col in nearby_venues3.columns]





print('{} venues for Oakland, California were returned by Foursquare.'.format(nearby_venues1.shape[0]))
print()
print('{} venues for Emeryville, California were returned by Foursquare.'.format(nearby_venues2.shape[0]))
print()
print('{} venues for San Diego, California were returned by Foursquare.'.format(nearby_venues3.shape[0]))

100 venues for Oakland, California were returned by Foursquare.

100 venues for Emeryville, California were returned by Foursquare.

100 venues for San Diego, California were returned by Foursquare.


In [5]:
# add locations data to the data sets of each city

nearby_venues1['city'] = 'Oakland'
nearby_venues2['city'] = 'Emeryville'
nearby_venues3['city'] = 'San Diego'

In [6]:
# combine the three cities into one data set

nearby_venues = nearby_venues1.copy()
nearby_venues = nearby_venues.append(nearby_venues2)
nearby_venues = nearby_venues.append(nearby_venues3)

In [7]:
nearby_venues

name  \
0                                          Oaklandish   
1                  Golden Lotus Vegetarian Restaurant   
2                                         Cape & Cowl   
3                                           Bar Shiru   
4                             Lucky Duck Bicycle Cafe   
5                                      Cafe Van Kleef   
6                                            Abura-Ya   
7                                 Woods Bar & Brewery   
8                                         Fox Theater   
9                        Nature Vegetarian Restaurant   
10                            Tay Ho Restaurant & Bar   
11                                Beauty’s Bagel Shop   
12                            The Cook And Her Farmer   
13                                     Binh Minh Quan   
14                                 Oakland Ice Center   
15                                         Umami Mart   
16                                 Blue Bottle Coffee   
17                                              Drexl   
18                                     Plenty Oakland   
19                                            Cosecha   
20                                       The Trappist   
21                                    The Sweet Booth   
22                         Kamdesh Afghan Kabob House   
23                                     Rio California   
24                          Golden Peacock Restaurant   
25                                       Miss Ollie's   
26                                         Sweetgreen   
27                                             Delage   
28                                 Cookiebar Creamery   
29                              Napoleon Super Bakery   
30                               Somar Bar and Lounge   
31                                      The Punchdown   
32                            Rosamunde Sausage Grill   
33                                       Belly Uptown   
34                                      First Edition   
35                        Great Western Power Company   
36                                          Shan Dong   
37                          Molcajete Cocina Mexicana   
38                                     Fresh and Best   
39                             Awaken Cafe & Roasting   
40                                Aroma Bakery & Cafe   
41                                   District Oakland   
42                        i-Tea Bubble Tea & Smoothie   
43                               Roam Artisan Burgers   
44                                      Swan's Market   
45                Ratto's International Market & Deli   
46                                       The Kon-Tiki   
47                                          The Hatch   
48                              Gyu-Kaku Japanese BBQ   
49                         Hopscotch Restaurant & Bar   
50                                  Paramount Theatre   
51                                          Spices! 3   
52                                     The New Parish   
53                                             Duende   
54                                        Mad Oak Bar   
55                                        Jamba Juice   
56                                       The Port Bar   
57                                          Radio Bar   
58                        La Borinqueña Mex-icatessen   
59                                             One Zo   
60                                         Burma Bear   
61                                         Parliament   
62                                 Blue Bottle Coffee   
63                                   Ike's Sandwiches   
64                                  Wooden Table Cafe   
65                                          Caffe 817   
66                                 Gum Kuo Restaurant   
67                                         Battambang   
68                   Original Pattern Brewing Company   
69                                        Falafel Boy   
70

In [8]:
nearby_venues['categories'].unique()

array(['Clothing Store', 'Vegetarian / Vegan Restaurant', 'Comic Shop',
       'Bar', 'Café', 'Japanese Restaurant', 'Brewery', 'Music Venue',
       'Vietnamese Restaurant', 'Bagel Shop', 'Seafood Restaurant',
       'Skating Rink', 'Kitchen Supply Store', 'Coffee Shop',
       'Mexican Restaurant', 'Beer Bar', 'Dessert Shop',
       'Afghan Restaurant', 'Brazilian Restaurant', 'Chinese Restaurant',
       'Caribbean Restaurant', 'Salad Place', 'Ice Cream Shop', 'Bakery',
       'Lounge', 'Wine Bar', 'Sausage Shop', 'Taco Place', 'Cocktail Bar',
       'Climbing Gym', 'Sandwich Place', 'Bubble Tea Shop',
       'Burger Joint', 'Food Court', 'Tiki Bar',
       'New American Restaurant', 'Nightclub', 'Tapas Restaurant',
       'Juice Bar', 'Gay Bar', 'Burmese Restaurant',
       'Cambodian Restaurant', 'Falafel Restaurant', 'Beer Garden',
       'Museum', 'Thai Restaurant', 'Indian Restaurant',
       'Street Food Gathering', 'ATM', 'Farmers Market',
       'Southern / Soul Food Restaur

#### **Data Cleaning**

In [9]:
# check list and manually remove all non-restaurant data

nearby_venues['categories'].unique()

removal_list = ['Clothing Store','Bar','Brewery', 
                'Comic Shop', 'Yoga Studio','Café', 
                'Coffee Shop', 'Tiki Bar', 'Music Venue', 
                'Wine Bar',  'Cocktail Bar', 'Dance Studio', 
                'Gym / Fitness Center','Beer Bar', 
                'Bubble Tea Shop', 'Nightclub', 'Food Court', 
                'Ice Cream Shop', 'Cupcake Shop', 'Skating Rink', 
                'Dessert Shop', 'Climbing Gym', 'Bakery', 
                'Farmers Market', 'Gay Bar','Beer Garden',
                'Tea Room','Arts & Crafts Store', 'Grocery Store', 
                'Sports Bar', 'Museum', 'Street Food Gathering', 
                'Library', 'Skate Park', 'Movie Theater','Park', 
                'Gym', 'Stadium', 'Furniture / Home Store', 'Discount Store', 
                'Playground', 'Cosmetics Shop', 'Casino', 
                'Pet Store','Electronics Store', 'Snack Place',
                'Salon / Barbershop', 'Shopping Plaza', 'Deli / Bodega', 
                'Candy Store', 'Liquor Store', 'Hotel', 
                'Shoe Store', 'Bookstore', 'Shopping Mall', 
                'Dive Bar', 'Video Game Store', 'Pharmacy', 
                'Accessories Store', 'Lingerie Store', 'Mobile Phone Shop', 
                'Pool Hall', 'Juice Bar', 'Kids Store', 
                'Supplement Shop', 'Big Box Store', 'Mattress Store', 
                'Hardware Store', 'Paper / Office Supplies Store', 'Theater', 
                'Business Service', 'Donut Shop', 'Beer Store', 
                'Lounge', 'Health Food Store', 'Pedestrian Plaza', 
                'Hookah Bar', 'Concert Hall', 'Chocolate Shop', 
                'Hostel', 'Convenience Store', 'Pub', 
                'Plaza', 'Comedy Club', 'Speakeasy', 
                'Tattoo Parlor', 'Massage Studio']

nearby_venues = nearby_venues[~nearby_venues['categories'].isin(removal_list)]

nearby_venues['categories'].unique().tolist()

['Vegetarian / Vegan Restaurant',
 'Japanese Restaurant',
 'Vietnamese Restaurant',
 'Bagel Shop',
 'Seafood Restaurant',
 'Kitchen Supply Store',
 'Mexican Restaurant',
 'Afghan Restaurant',
 'Brazilian Restaurant',
 'Chinese Restaurant',
 'Caribbean Restaurant',
 'Salad Place',
 'Sausage Shop',
 'Taco Place',
 'Sandwich Place',
 'Burger Joint',
 'New American Restaurant',
 'Tapas Restaurant',
 'Burmese Restaurant',
 'Cambodian Restaurant',
 'Falafel Restaurant',
 'Thai Restaurant',
 'Indian Restaurant',
 'ATM',
 'Southern / Soul Food Restaurant',
 'Breakfast Spot',
 'Ramen Restaurant',
 'Pizza Place',
 'Mediterranean Restaurant',
 'Diner',
 'Scandinavian Restaurant',
 'Wings Joint',
 'Asian Restaurant',
 'Burrito Place',
 'Fast Food Restaurant',
 'Sushi Restaurant',
 'Fried Chicken Joint',
 'Garden Center',
 'Food Truck',
 'American Restaurant',
 'Fondue Restaurant',
 'Theme Restaurant',
 'Italian Restaurant',
 'Hot Dog Joint',
 'Roof Deck',
 'Gastropub',
 'Neighborhood',
 'Steakhous

#### **DataFrame Creation**

In [10]:
# set up to pull the likes from the API based on venue ID

url_list = []
like_list = []
json_list = []

for i in list(nearby_venues.id):
    venue_url = 'https://api.foursquare.com/v2/venues/{}/likes?client_id={}&client_secret={}&v={}'.format(i, CLIENT_ID, CLIENT_SECRET, VERSION)
    url_list.append(venue_url)
for link in url_list:
    result = requests.get(link).json()
    likes = result['response']['likes']['count']
    like_list.append(likes)
print(like_list)


nearby_venues['likes'] = like_list
nearby_venues.head()

[77, 64, 33, 72, 22, 63, 50, 115, 199, 66, 40, 43, 100, 0, 51, 176, 182, 367, 33, 15, 25, 24, 39, 0, 237, 91, 171, 4, 27, 226, 76, 46, 9, 99, 13, 38, 261, 50, 1, 42, 13, 247, 122, 66, 35, 56, 36, 136, 132, 332, 24, 4, 1, 16, 36, 18, 54, 0, 31, 60, 64, 17, 92, 22, 68, 4, 17, 15, 7, 65, 31, 3, 5, 3, 7, 4, 156, 75, 33, 81, 18, 131, 171, 26, 105, 59, 41, 24, 22, 21, 31, 30, 2, 31, 35, 299, 61, 142, 133, 97, 12, 222, 31, 480, 104, 23, 187, 152, 34, 19, 140, 122, 325, 92, 54, 51, 13, 10, 488, 156, 10, 18]


name                     categories  \
1   Golden Lotus Vegetarian Restaurant  Vegetarian / Vegan Restaurant   
6                             Abura-Ya            Japanese Restaurant   
9         Nature Vegetarian Restaurant  Vegetarian / Vegan Restaurant   
10             Tay Ho Restaurant & Bar          Vietnamese Restaurant   
11                 Beauty’s Bagel Shop                     Bagel Shop   

          lat         lng                        id     city  likes  
1   37.803290 -122.270473  49cebb1bf964a520785a1fe3  Oakland     77  
6   37.805959 -122.267693  539a69a7498ee67090b2b285  Oakland     64  
9   37.802157 -122.270983  4f52deb2e4b0ac6d0c91df05  Oakland     33  
10  37.802062 -122.269573  4c8b16ec52a98cfad73533e9  Oakland     72  
11  37.806082 -122.268356  5bd0959cf1fdaf002ce03e11  Oakland     22

In [11]:
nearby_venues.count()

name          122
categories    122
lat           122
lng           122
id            122
city          122
likes         122
dtype: int64

In [12]:
# this is really the raw dataset now so let us rename it something more appropriate

raw_dataset = nearby_venues
raw_dataset.head()

name                     categories  \
1   Golden Lotus Vegetarian Restaurant  Vegetarian / Vegan Restaurant   
6                             Abura-Ya            Japanese Restaurant   
9         Nature Vegetarian Restaurant  Vegetarian / Vegan Restaurant   
10             Tay Ho Restaurant & Bar          Vietnamese Restaurant   
11                 Beauty’s Bagel Shop                     Bagel Shop   

          lat         lng                        id     city  likes  
1   37.803290 -122.270473  49cebb1bf964a520785a1fe3  Oakland     77  
6   37.805959 -122.267693  539a69a7498ee67090b2b285  Oakland     64  
9   37.802157 -122.270983  4f52deb2e4b0ac6d0c91df05  Oakland     33  
10  37.802062 -122.269573  4c8b16ec52a98cfad73533e9  Oakland     72  
11  37.806082 -122.268356  5bd0959cf1fdaf002ce03e11  Oakland     22

#### **Data Preperation for Machine Learning**

In [13]:
# inspecting the raw dataset shows that there may be too many different types of cuisines

raw_dataset['categories'].unique().tolist()

['Vegetarian / Vegan Restaurant',
 'Japanese Restaurant',
 'Vietnamese Restaurant',
 'Bagel Shop',
 'Seafood Restaurant',
 'Kitchen Supply Store',
 'Mexican Restaurant',
 'Afghan Restaurant',
 'Brazilian Restaurant',
 'Chinese Restaurant',
 'Caribbean Restaurant',
 'Salad Place',
 'Sausage Shop',
 'Taco Place',
 'Sandwich Place',
 'Burger Joint',
 'New American Restaurant',
 'Tapas Restaurant',
 'Burmese Restaurant',
 'Cambodian Restaurant',
 'Falafel Restaurant',
 'Thai Restaurant',
 'Indian Restaurant',
 'ATM',
 'Southern / Soul Food Restaurant',
 'Breakfast Spot',
 'Ramen Restaurant',
 'Pizza Place',
 'Mediterranean Restaurant',
 'Diner',
 'Scandinavian Restaurant',
 'Wings Joint',
 'Asian Restaurant',
 'Burrito Place',
 'Fast Food Restaurant',
 'Sushi Restaurant',
 'Fried Chicken Joint',
 'Garden Center',
 'Food Truck',
 'American Restaurant',
 'Fondue Restaurant',
 'Theme Restaurant',
 'Italian Restaurant',
 'Hot Dog Joint',
 'Roof Deck',
 'Gastropub',
 'Neighborhood',
 'Steakhous

In [14]:
# we can group some cuisines together to make a better categorical variable

european = ['Mediterranean Restaurant', 'Scandinavian Restaurant', 'Pizza Place',
       'French Restaurant', 'Falafel Restaurant', 'Italian Restaurant',
       'Turkish Restaurant']

latin = ['Mexican Restaurant', 'Taco Place', 'Brazilian Restaurant', 
          'Burrito Place']

asian = ['Japanese Restaurant', 'Vietnamese Restaurant', 'Chinese Restaurant',
         'Hot Dog Joint', 'Hotpot Restaurant', 'Indian Restaurant',
         'Thai Restaurant', 'Dumpling Restaurant', 'Dim Sum Restaurant',
         'Asian Restaurant', 'Filipino Restaurant', 'Sushi Restaurant',
         'Ramen Restaurant']

american = ['Vegetarian / Vegan Restaurant', 'Seafood Restaurant', 'Caribbean Restaurant',
           'Burger Joint', 'American Restaurant', 'New American Restaurant',
            'Southern / Soul Food Restaurant', 'Diner']

casual = ['Bagel Shop', 'Sandwich Place', 'Fried Chicken Joint', 
          'Breakfast Spot', 'Wings Joint', 'Fast Food Restaurant',
          'Theme Restaurant']

def conditions(s):
    if s['categories'] in european:
        return 'european'
    if s['categories'] in latin:
        return 'latin'
    if s['categories'] in asian:
        return 'asian'
    if s['categories'] in american:
        return 'american'
    if s['categories'] in casual:
        return 'casual'

raw_dataset['categories_classified'] = raw_dataset.apply(conditions, axis=1)
raw_dataset

name                       categories  \
1         Golden Lotus Vegetarian Restaurant    Vegetarian / Vegan Restaurant   
6                                   Abura-Ya              Japanese Restaurant   
9               Nature Vegetarian Restaurant    Vegetarian / Vegan Restaurant   
10                   Tay Ho Restaurant & Bar            Vietnamese Restaurant   
11                       Beauty’s Bagel Shop                       Bagel Shop   
12                   The Cook And Her Farmer               Seafood Restaurant   
13                            Binh Minh Quan            Vietnamese Restaurant   
15                                Umami Mart             Kitchen Supply Store   
19                                   Cosecha               Mexican Restaurant   
22                Kamdesh Afghan Kabob House                Afghan Restaurant   
23                            Rio California             Brazilian Restaurant   
24                 Golden Peacock Restaurant               Chinese Restaurant   
25                              Miss Ollie's             Caribbean Restaurant   
26                                Sweetgreen                      Salad Place   
27                                    Delage              Japanese Restaurant   
32                   Rosamunde Sausage Grill                     Sausage Shop   
33                              Belly Uptown                       Taco Place   
36                                 Shan Dong               Chinese Restaurant   
37                 Molcajete Cocina Mexicana               Mexican Restaurant   
38                            Fresh and Best            Vietnamese Restaurant   
40                       Aroma Bakery & Cafe                   Sandwich Place   
43                      Roam Artisan Burgers                     Burger Joint   
45       Ratto's International Market & Deli                   Sandwich Place   
48                     Gyu-Kaku Japanese BBQ              Japanese Restaurant   
49                Hopscotch Restaurant & Bar          New American Restaurant   
51                                 Spices! 3               Chinese Restaurant   
53                                    Duende                 Tapas Restaurant   
58               La Borinqueña Mex-icatessen               Mexican Restaurant   
60                                Burma Bear               Burmese Restaurant   
63                          Ike's Sandwiches                   Sandwich Place   
66                        Gum Kuo Restaurant               Chinese Restaurant   
67                                Battambang             Cambodian Restaurant   
69                               Falafel Boy               Falafel Restaurant   
70                                    b-Dama              Japanese Restaurant   
73                   Pintoh Thai Street Food                  Thai Restaurant   
76                              Curry Up Now                Indian Restaurant   
80                                      Xolo               Mexican Restaurant   
81                                    Phở 84            Vietnamese Restaurant   
83                             U.S. Bank ATM                              ATM   
84                     Vien Huong Restaurant            Vietnamese Restaurant   
88                       Brown Sugar Kitchen  Southern / Soul Food Restaurant   
89                                TrueBurger                     Burger Joint   
90                       Kingston 11 Cuisine             Caribbean Restaurant   
91                             The Gastropig                   Breakfast Spot   
94        Bay Fung Tong Tea House Restaurant               Chinese Restaurant   
96                 New Gold Medal Restaurant               Chinese Restaurant   
99                                   Shinmai                 Ramen Restaurant   
5                Arizmendi Bakery & Pizzeria                      Pizza Place   
6                               Wally's Cafe         Mediterranean Rest

In [15]:
# double check to make sure categories_classified has been created correctly

pd.crosstab(index = raw_dataset["categories_classified"],
            columns="count")

col_0                  count
categories_classified       
american                  24
asian                     28
casual                    16
european                  19
latin                     18

In [16]:
raw_dataset['likes'].mean()

79.15573770491804

In [17]:
# create a function to bin for us

def rankings(df):
    
    if df['likes'] <= 60:
        return 3
    
    elif df['likes'] <= 100:
        return 2
    
    elif df['likes'] > 100:
        return 1

In [18]:
# apply rankings function to dataset

raw_dataset['ranking'] = raw_dataset.apply(rankings, axis=1)
raw_dataset

name                       categories  \
1         Golden Lotus Vegetarian Restaurant    Vegetarian / Vegan Restaurant   
6                                   Abura-Ya              Japanese Restaurant   
9               Nature Vegetarian Restaurant    Vegetarian / Vegan Restaurant   
10                   Tay Ho Restaurant & Bar            Vietnamese Restaurant   
11                       Beauty’s Bagel Shop                       Bagel Shop   
12                   The Cook And Her Farmer               Seafood Restaurant   
13                            Binh Minh Quan            Vietnamese Restaurant   
15                                Umami Mart             Kitchen Supply Store   
19                                   Cosecha               Mexican Restaurant   
22                Kamdesh Afghan Kabob House                Afghan Restaurant   
23                            Rio California             Brazilian Restaurant   
24                 Golden Peacock Restaurant               Chinese Restaurant   
25                              Miss Ollie's             Caribbean Restaurant   
26                                Sweetgreen                      Salad Place   
27                                    Delage              Japanese Restaurant   
32                   Rosamunde Sausage Grill                     Sausage Shop   
33                              Belly Uptown                       Taco Place   
36                                 Shan Dong               Chinese Restaurant   
37                 Molcajete Cocina Mexicana               Mexican Restaurant   
38                            Fresh and Best            Vietnamese Restaurant   
40                       Aroma Bakery & Cafe                   Sandwich Place   
43                      Roam Artisan Burgers                     Burger Joint   
45       Ratto's International Market & Deli                   Sandwich Place   
48                     Gyu-Kaku Japanese BBQ              Japanese Restaurant   
49                Hopscotch Restaurant & Bar          New American Restaurant   
51                                 Spices! 3               Chinese Restaurant   
53                                    Duende                 Tapas Restaurant   
58               La Borinqueña Mex-icatessen               Mexican Restaurant   
60                                Burma Bear               Burmese Restaurant   
63                          Ike's Sandwiches                   Sandwich Place   
66                        Gum Kuo Restaurant               Chinese Restaurant   
67                                Battambang             Cambodian Restaurant   
69                               Falafel Boy               Falafel Restaurant   
70                                    b-Dama              Japanese Restaurant   
73                   Pintoh Thai Street Food                  Thai Restaurant   
76                              Curry Up Now                Indian Restaurant   
80                                      Xolo               Mexican Restaurant   
81                                    Phở 84            Vietnamese Restaurant   
83                             U.S. Bank ATM                              ATM   
84                     Vien Huong Restaurant            Vietnamese Restaurant   
88                       Brown Sugar Kitchen  Southern / Soul Food Restaurant   
89                                TrueBurger                     Burger Joint   
90                       Kingston 11 Cuisine             Caribbean Restaurant   
91                             The Gastropig                   Breakfast Spot   
94        Bay Fung Tong Tea House Restaurant               Chinese Restaurant   
96                 New Gold Medal Restaurant               Chinese Restaurant   
99                                   Shinmai                 Ramen Restaurant   
5                Arizmendi Bakery & Pizzeria                      Pizza Place   
6                               Wally's Cafe         Mediterranean Rest

#### **Machine Learning|Linear Regression**

In [19]:
# create dummies for linear regression modelling

# one hot encoding
reg_dataset = pd.get_dummies(raw_dataset[['categories_classified', 
                                          'city',]], 
                               prefix="", 
                               prefix_sep="")

# add name, ranking, and likes columns back to dataframe
reg_dataset['ranking'] = raw_dataset['ranking']
reg_dataset['likes'] = raw_dataset['likes']
reg_dataset['name'] = raw_dataset['name']

# move name column to the first column
reg_columns = [reg_dataset.columns[-1]] + list(reg_dataset.columns[:-1])
reg_dataset = reg_dataset[reg_columns]


reg_dataset.head()

name  american  asian  casual  european  \
1   Golden Lotus Vegetarian Restaurant         1      0       0         0   
6                             Abura-Ya         0      1       0         0   
9         Nature Vegetarian Restaurant         1      0       0         0   
10             Tay Ho Restaurant & Bar         0      1       0         0   
11                 Beauty’s Bagel Shop         0      0       1         0   

    latin  Emeryville  Oakland  San Diego  ranking  likes  
1       0           0        1          0        2     77  
6       0           0        1          0        2     64  
9       0           0        1          0        3     33  
10      0           0        1          0        2     72  
11      0           0        1          0        3     22

In [20]:
# Multiple Linear Regression

msk = np.random.rand(len(reg_dataset)) < 0.8
train = reg_dataset[msk]
test = reg_dataset[~msk]

regr = linear_model.LinearRegression()
x = np.asanyarray(train[['american', 'asian', 'casual',
                         'european', 'latin', 'Oakland', 
                         'Emeryville', 'San Diego']])

y = np.asanyarray(train[['likes']])
regr.fit (x, y)

# The coefficients

print ('Coefficients: ', regr.coef_)

Coefficients:  [[ 36.13906882 -19.1818694  -40.20189814 -26.58774278  19.74871495
    3.0707768  -29.41123248  26.34045568]]


In [21]:
# Multiple Linear Regression Prediction Capabilities

y_hat= regr.predict(test[['american', 'asian', 'casual',
                         'european', 'latin', 'Oakland', 
                         'Emeryville', 'San Diego']])

x = np.asanyarray(test[['american', 'asian', 'casual',
                         'european', 'latin', 'Oakland', 
                         'Emeryville', 'San Diego']])

y = np.asanyarray(test[['likes']])
print("Residual sum of squares: %.2f"
      % np.mean((y_hat - y) ** 2))

# Explained variance score: 1 is perfect prediction
print('Variance score: %.2f' % regr.score(x, y))

Residual sum of squares: 4647.90
Variance score: -0.04


#### **Machine Learning|Logistic Regression**

In [22]:
# Multinomial Ordinal Logistic Regression

x_train = np.asanyarray(train[['american', 'asian', 'casual',
                         'european', 'latin', 'Oakland', 
                         'Emeryville', 'San Diego']])

y_train = np.asanyarray(train['ranking'])

x_test = np.asanyarray(test[['american', 'asian', 'casual',
                         'european', 'latin', 'Oakland', 
                         'Emeryville', 'San Diego']])

y_test = np.asanyarray(test['ranking'])


mul_ordinal = linear_model.LogisticRegression(multi_class='multinomial',
                                              solver='newton-cg',
                                              fit_intercept=True).fit(x_train,
                                                                      y_train)

mul_ordinal

coef = mul_ordinal.coef_[0]
print (coef)

[ 0.30909275 -0.40682944 -0.35894913  0.01936095  0.08634385  0.02030906
 -0.4892652   0.46896462]


In [23]:
# Multinomial Ordinal Logistic Regression Prediction Capabilities

yhat = mul_ordinal.predict(x_test)
yhat

yhat_prob = mul_ordinal.predict_proba(x_test)
yhat_prob


# average = None, average = 'micro', average = 'macro', or average = 'weighted'
jaccard_score(y_test, yhat, average='weighted')

0.4479338842975206

In [24]:
log_loss(y_test, yhat_prob)

0.8629705758012701

In [25]:
# Exploration of Coefficient Magnitudes of Full Dataset

x_all = np.asanyarray(reg_dataset[['american', 'asian', 'casual',
                                   'european', 'latin', 'Oakland', 
                                   'Emeryville', 'San Diego']])

y_all = np.asanyarray(reg_dataset['ranking'])



LR = linear_model.LogisticRegression(multi_class='multinomial',
                                            solver='newton-cg',
                                            fit_intercept=True).fit(x_all,
                                                                    y_all)

LR

coef = LR.coef_[0]
print(coef)

[ 0.25032732 -0.66672001 -0.39255731 -0.0185364   0.00421523  0.04814135
 -0.47629323  0.42818353]


In [26]:
print(classification_report(y_test, yhat))

              precision    recall  f1-score   support

           1       0.50      0.38      0.43         8
           2       0.00      0.00      0.00         4
           3       0.67      0.86      0.75        21

    accuracy                           0.64        33
   macro avg       0.39      0.41      0.39        33
weighted avg       0.55      0.64      0.58        33



## Results and Discussion <a name="results"></a>

A linear regression model was trained on a random subsample of 80% and then the other 20% was used for testing purposes. In order to evaluate if the model is reasonable, the residual sum of squares and variance score were both calculated (10159.10, 0.16). The variance score is quite low, which means that is not a good way of modeling our data. Therefore, we moved on to logistic regression for our analysis.

The multinomial ordinal logistic regression model was also trained on a random subsample of 80% and then tested on the remaining 20%. The jaccard score and log-loss were both calculated (26.19% and 1.278 respectively). Although the prediction is not promosing, a jaccard score of 26% is somewhat reasonable. The classification report is present in the analysis.

Given the modestly accurate ability of this mode, we have the ability to run the model on the complete dataset. The coefficients we got show that opening a restaurant in Emeryville, or serving cuisine that is asian, or casual, are negatively associated with likes.


Logistic regression presents a better fit for the data over linear regression. Using logistic regression we were able to obtain a Jaccard Score of 26.19%, which although not perfect, is more reasonable than the low variance score obtained from the linear regression. As stated before, please note that for the purposes of this project, we are assumming that likes are a good proxy for how well a new restaurant will do in terms of brand, image and by extension how well the restaurant will perform business-wise. Whether or not these assumptions hold up in a real-life scenario is up for discussion, but this project does contain limitations in scope due to the amount of data that can be fetched from the FourSquare API.

As such, to obtain insights into this data, we can proceed with breaking down the results of the logistic regression model. The results showed that the precision score for classifying whether the new restaurant would fall into classes 1, 2, or 3 (highest, medium, lowest) were 40%, 0%, and 50%. Therefore, the model is better at predicting if a restaurant will fall into the best or worst percentile of likes. This is good as we are mostly concerned with whether the restuarant will perform well or not so the high accuracy of predictions for the two extremum is a welcome feature. This allows us to fairly accurately predict the general performance of the business opportunity. Different binning methods for the classes were attempted, but the use of 3 bins by far yielded the best Jaccard Score.

Additionally, not only are we attempting to predict the general business performance but also pull insights to inform on business strategy. In this case strategy insight can be gleamed from the coefficient values from running the logistic regression on the full dataset. As such, we can see that opening a restaurant in Emeryville, or serving cuisine that is asian or casual in nature, are associated negatively with "likes." This suggests that the business opportunity should be opening a restaurant in either Oakland or San Diego, with a cuisine that is European, Latin, or American in nature would be the best approach for maximizing likes.

## Conclusion <a name="conclusion"></a>

In conclusion, after analyzing restaurant likes in California from the 300 restaurants, we can conclude that the approach to best take when looking to maximize business performance (as measured by 'likes') is to open a restaurant that is either European, Latino, or American and that opening the venue in either Oakland or San Diego rather than Emeryville would be the best approach. Additionally, the predictive capabilities of the logistic regression prediction model proved to be the most accurate for classifying whether a restaurant fell in either the best or worst classes when the data was binned into their 3 respective classes.